src/role.rs

```diff
@@ -1,88 +1,64 @@
-use crate::{Error, PoiseContext};
-use crate::config::CONFIG;
+use crate::{config::CONFIG, Error, PoiseContext};
 
-use std::collections::HashMap;
-use std::hash::Hash;
-use rand::thread_rng;
-use serenity::http::CacheHttp;
-use rand::seq::{IteratorRandom};
-use serenity::all::{Member, Permissions};
-use serenity::builder::EditRole;
-use serenity::prelude::Context;
+use rand::{seq::IteratorRandom, thread_rng};
+use serenity::{
+    all::{Member, Permissions},
+    builder::EditRole,
+    http::CacheHttp,
+    prelude::Context,
+};
+use std::{collections::HashMap, hash::Hash};
 
 fn rand_hash<K: Eq + Hash, V>(hash: &HashMap<K, V>) -> &K {
-    let key = hash.keys().choose(&mut thread_rng()).unwrap();
-
-    key
+    hash.keys().choose(&mut thread_rng()).unwrap()
 }
 
 pub async fn random_color(ctx: &Context, member: &Member) -> Result<(), Error> {
-    let choice = rand_hash(&CONFIG.colors);
-
     let guild = match member.guild_id.to_guild_cached(&ctx.cache) {
         Some(g) => g.clone(),
-        None => return Ok(())
+        None => return Ok(()),
     };
 
-    let r = EditRole::new().name(choice.as_str()).colour(CONFIG.colors[choice]).permissions(Permissions::empty());
-    let role_id = match guild.role_by_name(choice.as_str()) {
+    let choice = rand_hash(&CONFIG.colors);
+    let r = EditRole::new().name(choice).colour(CONFIG.colors[choice]).permissions(Permissions::empty());
```

Comment:
could we just call `process_color` with this `choice` here? the rest of the function looks identical to me.

Reply:
Context is actually different, as one is called in a slash-command, while the other is called from a member join event

Reply:
Looks like [Poise's context](https://docs.rs/poise/latest/poise/structs/enum.Context.html) is a thin wrapper around two different structs providing different wrappers around Serenity's. There's some shared interface that makes [`send_reply`](https://docs.rs/poise/latest/poise/reply/fn.send_reply.html) and [`say_reply`](https://docs.rs/poise/latest/poise/reply/fn.say_reply.html) look like a method-independent manner of responding in the context. Context agnostic rewrite seems doable?